File: app/src/main/java/org/wikipedia/readinglist/ReadingListTitleDialog.java
package org.wikipedia.readinglist;

import android.content.Context;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import org.apache.commons.lang3.StringUtils;
import org.wikipedia.R;
import org.wikipedia.views.TextInputDialog;

import java.util.List;

public final class ReadingListTitleDialog {

    public interface Callback {
        void onSuccess(@NonNull String text, @NonNull String description);
    }

    public static TextInputDialog readingListTitleDialog(@NonNull Context context,
                                                         @NonNull final String title,
                                                         @Nullable final String description,
                                                         @NonNull final List<String> otherTitles,
                                                         @Nullable final Callback callback) {
        return TextInputDialog.newInstance(context, true,
                new TextInputDialog.Callback() {
                    @Override
                    public void onShow(@NonNull TextInputDialog dialog) {
                        dialog.setHint(R.string.reading_list_name_hint);
                        dialog.setSecondaryHint(R.string.reading_list_description_hint);
                        dialog.setText(title);
                        dialog.setSecondaryText(StringUtils.defaultString(description));
                    }

                    @Override
                    public void onTextChanged(@NonNull CharSequence text, @NonNull TextInputDialog dialog) {
                        String title = text.toString().trim();
                        if (StringUtils.isEmpty(title)) {
                            dialog.setError(null);
                            dialog.setPositiveButtonEnabled(false);
                        } else if (otherTitles.contains(title)) {
                            dialog.setError(dialog.getContext().getString(R.string.reading_list_title_exists, title));
                            dialog.setPositiveButtonEnabled(false);
                        } else {
                            dialog.setError(null);
                            dialog.setPositiveButtonEnabled(true);
                        }
                    }

                    @Override
                    public void onSuccess(@NonNull CharSequence text, @NonNull CharSequence secondaryText) {
                        if (callback != null) {
                            callback.onSuccess(text.toString().trim(), secondaryText.toString().trim());
                        }
                    }

                    @Override
                    public void onCancel() {
                    }
                });
    }

    private ReadingListTitleDialog() {
    }
}

File: app/src/main/java/org/wikipedia/readinglist/ReadingListTitleDialog.kt
package org.wikipedia.readinglist

import android.app.Activity
import org.wikipedia.R
import org.wikipedia.views.TextInputDialog

object ReadingListTitleDialog {
    fun interface Callback {
        fun onSuccess(text: String, description: String)
    }

    @JvmStatic
    fun readingListTitleDialog(activity: Activity,
                               title: String,
                               description: String?,
                               otherTitles: List<String?>,
                               callback: Callback?): TextInputDialog {
        return TextInputDialog(activity).let { textInputDialog ->
            textInputDialog.callback = object : TextInputDialog.Callback {
                override fun onShow(dialog: TextInputDialog) {
                    dialog.setHint(R.string.reading_list_name_hint)
                    dialog.setSecondaryHint(R.string.reading_list_description_hint)
                    dialog.setText(title, true)
                    dialog.setSecondaryText(description.orEmpty())
                }

                override fun onTextChanged(text: CharSequence, dialog: TextInputDialog) {
                    text.toString().trim().let {
                        when {
                            it.isEmpty() -> {
                                dialog.setError(null)
                                dialog.setPositiveButtonEnabled(false)
                            }
                            otherTitles.contains(it) -> {
                                dialog.setError(dialog.context.getString(R.string.reading_list_title_exists, it))
                                dialog.setPositiveButtonEnabled(false)
                            }
                            else -> {
                                dialog.setError(null)
                                dialog.setPositiveButtonEnabled(true)
                            }
                        }
                    }
                }

                override fun onSuccess(text: CharSequence, secondaryText: CharSequence) {
                    callback?.onSuccess(text.toString().trim(), secondaryText.toString().trim())
                }

                override fun onCancel() {}
            }
            textInputDialog.showSecondaryText(true)
        }
    }
}
